### ISSUE TITLE
Difference between tutorial_mnist_mlp_static.py and tutorial_mnist_mlp_static_2.py

### ISSUE THREAD

**Issue by 0xtyls** [tensorlayer/tensorlayer#1054]
I understand that these two python files show two different methods to construct a model.  The original n_epoch is 500 which works perfect for both python files. But if I change n_epoch to 20, only tutorial_mnist_mlp_static.py can achieve a high test accuracy (~0.97). The other file tutorial_mnist_mlp_static_2.py only get 0.47. 

The models built from these two files looks the same for me (the same number of layers, the same n_units for each layer...). So why? What is the difference between these two methods?

**Comment by Laicheng0830** [comment]
I changed the n_epoch of tutorial_mnist_mlp_static_2.py to 20 and the test accuracy is ~0.97, no repetition of your results.

can you give some reproducible code?

**Comment by 0xtyls** [comment]
`import time
import numpy as np
import tensorflow as tf

import tensorlayer as tl
from tensorlayer.layers import Dense, Dropout, Input
from tensorlayer.models import Model

## enable debug logging
tl.logging.set_verbosity(tl.logging.DEBUG)

## prepare MNIST data
X_train, y_train, X_val, y_val, X_test, y_test = tl.files.load_mnist_dataset(shape=(-1, 784))

## define the network
# the softmax is implemented internally in tl.cost.cross_entropy(y, y_) to
# speed up computation, so we use identity here.
# see tf.nn.sparse_softmax_cross_entropy_with_logits()


def hidden_model(inputs_shape):
    ni = Input(inputs_shape)
    nn = Dropout(keep=0.8)(ni)
    nn = Dense(n_units=800, act=tf.nn.relu)(nn)
    nn = Dropout(keep=0.8)(nn)
    nn = Dense(n_units=800, act=tf.nn.relu)(nn)

    return Model(inputs=ni, outputs=nn, name="mlp_hidden")


def get_model(inputs_shape, hmodel):
    hidden = hmodel.as_layer()
    ni = Input(inputs_shape)
    nn = hidden(ni)
    nn = Dropout(keep=0.8)(nn)
    nn = Dense(n_units=10, act=tf.nn.relu)(nn)

    return Model(inputs=ni, outputs=nn, name="mlp")


MLP_hidden = hidden_model([None, 784])
MLP = get_model([None, 784], MLP_hidden)
# MLP.print_layers()
# MLP.print_weights()

## start training
n_epoch = 20
batch_size = 500
print_freq = 5
train_weights = MLP.trainable_weights
optimizer = tf.optimizers.Adam(lr=0.0001)

## the following code can help you understand SGD deeply
for epoch in range(n_epoch):  ## iterate the dataset n_epoch times
    start_time = time.time()
    ## iterate over the entire training set once (shuffle the data via training)
    for X_batch, y_batch in tl.iterate.minibatches(X_train, y_train, batch_size, shuffle=True):
        MLP.train()  # enable dropout
        with tf.GradientTape() as tape:
            ## compute outputs
            _logits = MLP(X_batch)  # alternatively, you can use MLP(x, is_train=True) and remove MLP.train()
            ## compute loss and update model
            _loss = tl.cost.cross_entropy(_logits, y_batch, name='train_loss')
        grad = tape.gradient(_loss, train_weights)
        optimizer.apply_gradients(zip(grad, train_weights))

    ## use training and evaluation sets to evaluate the model every print_freq epoch
    if epoch + 1 == 1 or (epoch + 1) % print_freq == 0:
        MLP.eval()  # disable dropout
        print("Epoch {} of {} took {}".format(epoch + 1, n_epoch, time.time() - start_time))
        train_loss, train_acc, n_iter = 0, 0, 0
        for X_batch, y_batch in tl.iterate.minibatches(X_train, y_train, batch_size, shuffle=False):

            _logits = MLP(X_batch)  # alternatively, you can use MLP(x, is_train=False) and remove MLP.eval()
            train_loss += tl.cost.cross_entropy(_logits, y_batch, name='eval_loss')
            train_acc += np.mean(np.equal(np.argmax(_logits, 1), y_batch))
            n_iter += 1
        print("   train loss: {}".format(train_loss / n_iter))
        print("   train acc:  {}".format(train_acc / n_iter))
        val_loss, val_acc, n_iter = 0, 0, 0
        for X_batch, y_batch in tl.iterate.minibatches(X_val, y_val, batch_size, shuffle=False):
            _logits = MLP(X_batch)  # is_train=False, disable dropout
            val_loss += tl.cost.cross_entropy(_logits, y_batch, name='eval_loss')
            val_acc += np.mean(np.equal(np.argmax(_logits, 1), y_batch))
            n_iter += 1
        print("   val loss: {}".format(val_loss / n_iter))
        print("   val acc:  {}".format(val_acc / n_iter))

## use testing data to evaluate the model
MLP.eval()
test_loss, test_acc, n_iter = 0, 0, 0
for X_batch, y_batch in tl.iterate.minibatches(X_test, y_test, batch_size, shuffle=False):
    _logits = MLP(X_batch)
    test_loss += tl.cost.cross_entropy(_logits, y_batch, name='test_loss')
    test_acc += np.mean(np.equal(np.argmax(_logits, 1), y_batch))
    n_iter += 1
print("   test loss: {}".format(test_loss / n_iter))
print("   test acc:  {}".format(test_acc / n_iter))`

It's almost the same as tutorial_mnist_mlp_static_2.py except n_epoch is 20.

Here is the output:

`[TL] Load or Download MNIST > data/mnist
[TL] data/mnist/train-images-idx3-ubyte.gz
[TL] data/mnist/t10k-images-idx3-ubyte.gz
[TL] Input  _inputlayer_1: [None, 784]
2020-01-10 13:56:35.326152: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2020-01-10 13:56:35.334688: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 3100010000 Hz
2020-01-10 13:56:35.337981: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x3d85660 executing computations on platform Host. Devices:
2020-01-10 13:56:35.338013: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): Host, Default Version
[TL] Dropout dropout_1: keep: 0.800000 
[TL] Dense  dense_1: 800 relu
[TL] Dropout dropout_2: keep: 0.800000 
[TL] Dense  dense_2: 800 relu
[TL] ModelLayer modellayer_1 from Model: mlp_hidden
[TL] Input  _inputlayer_2: [None, 784]
[TL] Dropout dropout_3: keep: 0.800000 
[TL] Dense  dense_3: 10 relu
Epoch 1 of 20 took 1.862147569656372
   train loss: 2.316465139389038
   train acc:  0.1945600000000001
   val loss: 2.327671527862549
   val acc:  0.19480000000000003
Epoch 5 of 20 took 1.8243627548217773
   train loss: 2.039196491241455
   train acc:  0.2961999999999999
   val loss: 2.033693790435791
   val acc:  0.2982
Epoch 10 of 20 took 1.867743968963623
   train loss: 1.8629565238952637
   train acc:  0.3595800000000001
   val loss: 1.8483893871307373
   val acc:  0.36410000000000003
Epoch 15 of 20 took 1.9024789333343506
   train loss: 1.4744011163711548
   train acc:  0.4884799999999998
   val loss: 1.437483787536621
   val acc:  0.5021000000000001
Epoch 20 of 20 took 1.834416389465332
   train loss: 1.3005656003952026
   train acc:  0.5417399999999999
   val loss: 1.267709493637085
   val acc:  0.5511
   test loss: 1.2561590671539307
   test acc:  0.5582`

**Comment by 0xtyls** [comment]
```python
import time
import numpy as np
import tensorflow as tf

import tensorlayer as tl
from tensorlayer.layers import Dense, Dropout, Input
from tensorlayer.models import Model

tl.logging.set_verbosity(tl.logging.DEBUG)

X_train, y_train, X_val, y_val, X_test, y_test = tl.files.load_mnist_dataset(shape=(-1, 784))


def hidden_model(inputs_shape):
    ni = Input(inputs_shape)
    nn = Dropout(keep=0.8)(ni)
    nn = Dense(n_units=800, act=tf.nn.relu)(nn)
    nn = Dropout(keep=0.8)(nn)
    nn = Dense(n_units=800, act=tf.nn.relu)(nn)

    return Model(inputs=ni, outputs=nn, name="mlp_hidden")


def get_model(inputs_shape, hmodel):
    hidden = hmodel.as_layer()
    ni = Input(inputs_shape)
    nn = hidden(ni)
    nn = Dropout(keep=0.8)(nn)
    nn = Dense(n_units=10, act=tf.nn.relu)(nn)

    return Model(inputs=ni, outputs=nn, name="mlp")


MLP_hidden = hidden_model([None, 784])
MLP = get_model([None, 784], MLP_hidden)

n_epoch = 20
batch_size = 500
print_freq = 5
train_weights = MLP.trainable_weights
optimizer = tf.optimizers.Adam(lr=0.0001)

for epoch in range(n_epoch):
    start_time = time.time()
    for X_batch, y_batch in tl.iterate.minibatches(X_train, y_train, batch_size, shuffle=True):
        MLP.train()
        with tf.GradientTape() as tape:
            _logits = MLP(X_batch)
            _loss = tl.cost.cross_entropy(_logits, y_batch, name='train_loss')
        grad = tape.gradient(_loss, train_weights)
        optimizer.apply_gradients(zip(grad, train_weights))

    if epoch + 1 == 1 or (epoch + 1) % print_freq == 0:
        MLP.eval()
        print("Epoch {} of {} took {}".format(epoch + 1, n_epoch, time.time() - start_time))
        train_loss, train_acc, n_iter = 0, 0, 0
        for X_batch, y_batch in tl.iterate.minibatches(X_train, y_train, batch_size, shuffle=False):

            _logits = MLP(X_batch)
            train_loss += tl.cost.cross_entropy(_logits, y_batch, name='eval_loss')
            train_acc += np.mean(np.equal(np.argmax(_logits, 1), y_batch))
            n_iter += 1
        print("   train loss: {}".format(train_loss / n_iter))
        print("   train acc:  {}".format(train_acc / n_iter))
        val_loss, val_acc, n_iter = 0, 0, 0
        for X_batch, y_batch in tl.iterate.minibatches(X_val, y_val, batch_size, shuffle=False):
            _logits = MLP(X_batch)
            val_loss += tl.cost.cross_entropy(_logits, y_batch, name='eval_loss')
            val_acc += np.mean(np.equal(np.argmax(_logits, 1), y_batch))
            n_iter += 1
        print("   val loss: {}".format(val_loss / n_iter))
        print("   val acc:  {}".format(val_acc / n_iter))

MLP.eval()
test_loss, test_acc, n_iter = 0, 0, 0
for X_batch, y_batch in tl.iterate.minibatches(X_test, y_test, batch_size, shuffle=False):
    _logits = MLP(X_batch)
    test_loss += tl.cost.cross_entropy(_logits, y_batch, name='test_loss')
    test_acc += np.mean(np.equal(np.argmax(_logits, 1), y_batch))
    n_iter += 1
print("   test loss: {}".format(test_loss / n_iter))
print("   test acc:  {}".format(test_acc / n_iter))
```

It is almost the same as tutorial_mnist_mlp_static_2.py except n_epoch is 20. 

Here is the console output:

```bash
[TL] Load or Download MNIST > data/mnist
[TL] data/mnist/train-images-idx3-ubyte.gz
[TL] data/mnist/t10k-images-idx3-ubyte.gz
[TL] Input  _inputlayer_1: [None, 784]
2020-01-10 13:56:35.326152: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2020-01-10 13:56:35.334688: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 3100010000 Hz
2020-01-10 13:56:35.337981: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x3d85660 executing computations on platform Host. Devices:
2020-01-10 13:56:35.338013: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): Host, Default Version
[TL] Dropout dropout_1: keep: 0.800000 
[TL] Dense  dense_1: 800 relu
[TL] Dropout dropout_2: keep: 0.800000 
[TL] Dense  dense_2: 800 relu
[TL] ModelLayer modellayer_1 from Model: mlp_hidden
[TL] Input  _inputlayer_2: [None, 784]
[TL] Dropout dropout_3: keep: 0.800000 
[TL] Dense  dense_3: 10 relu
Epoch 1 of 20 took 1.862147569656372
   train loss: 2.316465139389038
   train acc:  0.1945600000000001
   val loss: 2.327671527862549
   val acc:  0.19480000000000003
Epoch 5 of 20 took 1.8243627548217773
   train loss: 2.039196491241455
   train acc:  0.2961999999999999
   val loss: 2.033693790435791
   val acc:  0.2982
Epoch 10 of 20 took 1.867743968963623
   train loss: 1.8629565238952637
   train acc:  0.3595800000000001
   val loss: 1.8483893871307373
   val acc:  0.36410000000000003
Epoch 15 of 20 took 1.9024789333343506
   train loss: 1.4744011163711548
   train acc:  0.4884799999999998
   val loss: 1.437483787536621
   val acc:  0.5021000000000001
Epoch 20 of 20 took 1.834416389465332
   train loss: 1.3005656003952026
   train acc:  0.5417399999999999
   val loss: 1.267709493637085
   val acc:  0.5511
   test loss: 1.2561590671539307
   test acc:  0.5582
```

**Comment by Laicheng0830** [comment]
You should upgrade tensorlayer2.1.0 to the tensorlayer2.2.1
